### PR TITLE
Add `rooms.bump_stamp` to Sliding Sync `/sync` for easier client-side sorting

### PR DESCRIPTION
**Pull request by MadLittleMods**
Add `rooms.bump_stamp` to Sliding Sync `/sync` for easier client-side sorting.

`bump_stamp` corresponds to the `stream_ordering` of the latest `DEFAULT_BUMP_EVENT_TYPES` in the room. This helps clients sort more readily without them needing to pull in a bunch of the timeline to determine the last activity. `bump_event_types` is a thing because for example, we don't want display name changes to mark the room as unread and bump it to the top. For encrypted rooms, we just have to consider any activity as a bump because we can't see the content and the client has to figure it out for themselves.

Outside of Synapse, `bump_stamp` is just a free-form counter so other implementations could use `received_ts`or `origin_server_ts` (see the [*Security considerations* section in MSC3575 about the potential pitfalls of using `origin_server_ts`](https://github.com/matrix-org/matrix-spec-proposals/blob/kegan/sync-v3/proposals/3575-sync.md#security-considerations)). It doesn't have any guarantee about always going up. In the Synapse case, it could go down if an event was redacted/removed (or purged in cases of retention policies).

In the future, we could add `bump_event_types` as [MSC3575](https://github.com/matrix-org/matrix-spec-proposals/pull/3575) mentions if people need to customize the event types.

---

In the Sliding Sync proxy, a similar [`timestamp` field was added](https://github.com/matrix-org/sliding-sync/pull/247) for the same purpose but the name is not obvious what it pertains to or what it's for.

The `timestamp` field was also added to Ruma in https://github.com/ruma/ruma/pull/1622

### Dev notes

[Discussion in an internal room](https://matrix.to/#/!cnVVNLKqgUzNTOFQkz:matrix.org/$UV-3yscsxX6q6mt3CzzyCRjvLgbV35ZgQuGf5UOCz4c?via=element.io&via=matrix.org&via=jki.re)

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/handlers/sliding_sync.py [hunk 1]
```diff
@@ -122,6 +134,62 @@ def copy_and_replace(self, **kwds: Any) -> "_RoomMembershipForUser":
...
+    """
+    Same as `_RoomMembershipForUser` but with an additional `bump_stamp` attribute.
+
+    Attributes:
+        event_id: The event ID of the membership event
+        event_pos: The stream position of the membership event
+        membership: The membership state of the user in the room
+        sender: The person who sent the membership event
+        newly_joined: Whether the user newly joined the room during the given token
+            range
```
Comment (erikjohnston) [inline synapse/handlers/sliding_sync.py:148]:
```suggestion
    Attributes:
```

I don't think we should double document these attributes?

**Review by erikjohnston — COMMENTED** [review]
Don't we need to add this to the API?

File: synapse/handlers/sliding_sync.py [hunk 2]
```diff
@@ -782,36 +867,49 @@ async def sort_rooms(
...
+                        room_for_user, bump_stamp=room_for_user.event_pos.stream
+                    )
+                )
 
         return sorted(
-            sync_room_map.items(),
+            sorted_sync_rooms,
             # Sort by the last activity (stream_ordering) in the room
-            key=lambda room_info: last_activity_in_room_map[room_info[0]],
+            key=lambda room_info: room_info.bump_stamp,
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync.py:912]: So clients receive all new events, we should still sort by the latest event. Clients can sort based on the `bump_stamp` field though.

File: synapse/api/constants.py [hunk 3]
```diff
@@ -128,9 +128,13 @@ class EventTypes:
     SpaceParent: Final = "m.space.parent"
 
     Reaction: Final = "m.reaction"
+    Sticker: Final = "m.sticker"
+    LocationShare: Final = "m.location"
```
Comment (MadLittleMods) [inline synapse/api/constants.py:132]: @MatMaul pointed out live location sharing is actually `m.beacon_info` ([MSC3489](https://github.com/matrix-org/matrix-spec-proposals/pull/3489))

`m.location` is just a normal `m.room.message`, https://spec.matrix.org/v1.10/client-server-api/#mlocation

Example:
```json
{
  "content": {
    "body": "Big Ben, London, UK",
    "geo_uri": "geo:51.5008,0.1247",
    "info": {
      "thumbnail_info": {
        "h": 300,
        "mimetype": "image/jpeg",
        "size": 46144,
        "w": 300
      },
      "thumbnail_url": "mxc://example.org/FHyPlCeYUSFFxlgbQYZmoEoe"
    },
    "msgtype": "m.location"
  },
  "event_id": "$143273582443PhrSn:example.org",
  "origin_server_ts": 1432735824653,
  "room_id": "!jEsUZKDJdhlrceRyVU:example.org",
  "sender": "@example:example.org",
  "type": "m.room.message",
  "unsigned": {
    "age": 1234
  }
}
```

File: synapse/handlers/sliding_sync.py (hunk 2 above)
Comment (MadLittleMods) [reply synapse/handlers/sliding_sync.py]: Reverted back to sorting by the latest event :white_check_mark: 

We still have the `bump_stamp` field though for clients to do what they want with.

File: synapse/api/constants.py (hunk 3 above)
Comment (MadLittleMods) [reply synapse/api/constants.py:132]: Updated to use `m.beacon_info` :white_check_mark:

File: tests/rest/client/test_sync.py [hunk 4]
```diff
@@ -1937,6 +1937,102 @@ def test_rooms_incremental_sync(self) -> None:
...
+                    "room_ids": [room_id1, room_id2],
+                }
+            ],
+            channel.json_body["lists"]["foo-list"],
+        )
+
+        # The `bump_stamp` for room1 should point at the latest message (not the
+        # reaction since it's not one of the `DEFAULT_BUMP_EVENT_TYPES`)
+        self.assertEqual(
+            channel.json_body["rooms"][room_id1]["bump_stamp"],
```
Comment (MadLittleMods) [inline tests/rest/client/test_sync.py:2024]: > Don't we need to add this to the API?

:clown_face: oops sorry for missing this :bow: 

Added to the API response and added a test for it now :white_check_mark:

File: synapse/handlers/sliding_sync.py [hunk 5]
```diff
@@ -1358,6 +1394,7 @@ async def get_room_sync_data(
             prev_batch=prev_batch_token,
             limited=limited,
             num_live=num_live,
+            bump_stamp=bump_stamp,
```
Comment (erikjohnston) [inline synapse/handlers/sliding_sync.py:1397]: I'm slightly dubious about giving back an opaque number rather than a timestamp, but we can iterate later.

**Review by erikjohnston — APPROVED** [review]

**Comment by MadLittleMods** [comment]
Thanks for the review @erikjohnston 🐏

**Pull request merged**